data/src/main/java/me/androidbox/data/worker_manager/SyncFullAgendaItemsWorker.kt

```diff
@@ -0,0 +1,127 @@
+import androidx.work.CoroutineWorker
+import androidx.work.WorkerParameters
+import dagger.assisted.Assisted
+import dagger.assisted.AssistedInject
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.supervisorScope
+import me.androidbox.data.local.dao.EventDao
+import me.androidbox.data.local.dao.ReminderDao
+import me.androidbox.data.local.dao.TaskDao
+import me.androidbox.data.local.entity.EventEntity
+import me.androidbox.data.local.entity.ReminderEntity
+import me.androidbox.data.local.entity.TaskEntity
+import me.androidbox.data.mapper.toAttendee
+import me.androidbox.data.remote.model.response.FullAgendaDto
+import me.androidbox.data.remote.network.agenda.AgendaService
+import me.androidbox.data.remote.util.CheckResult.checkResult
+import me.androidbox.data.worker_manager.constant.Constant.RETRY_COUNT
+
+@HiltWorker
+class SyncFullAgendaItemsWorker @AssistedInject constructor(
+    @Assisted private val context: Context,
+    @Assisted private val workerParameters: WorkerParameters,
+    private val agendaService: AgendaService,
+    private val eventDao: EventDao,
+    private val taskDao: TaskDao,
+    private val reminderDao: ReminderDao,
+) : CoroutineWorker(context, workerParameters) {
+
+    override suspend fun doWork(): Result {
+
+        val responseResult = checkResult {
+            agendaService.fullAgenda()
+        }
+
+        if(runAttemptCount > RETRY_COUNT) {
+            Result.failure()
+        }
+
+        val result = responseResult.fold(
+            onSuccess = { fullAgendaDto ->
+                insertAllAgendaItems(fullAgendaDto, eventDao, taskDao, reminderDao)
+                Result.success()
+            },
+            onFailure = {
+                Result.retry()
+            }
+        )
+
+        return result
+    }
+}
+
+private suspend fun insertAllAgendaItems(fullAgendaDto: FullAgendaDto, eventDao: EventDao, taskDao: TaskDao, reminderDao: ReminderDao) {
+    supervisorScope {
+        val eventJob = launch {
+            fullAgendaDto.events.mapAsync { eventDto ->
+                val eventEntity = EventEntity(
+                    id = eventDto.id,
+                    title = eventDto.title,
+                    description = eventDto.description,
+                    startDateTime = eventDto.from,
+                    endDateTime = eventDto.to,
+                    remindAt = eventDto.remindAt,
+                    eventCreatorId = eventDto.host,
+                    isUserEventCreator = eventDto.isUserEventCreator,
+                    photos = eventDto.photos.map { photoDto ->  photoDto.key },
+                    isGoing = true,
+                    attendees = eventDto.attendees.map { attendeeDto ->
+                        attendeeDto.toAttendee()
+                    }
+                )
+
+                eventDao.insertEvent(eventEntity)
```

Comment:
Instead of inserting one item at a time, you could have an "insertEvents" that accepts a list

So then you'd just:
```
val eventEntities = fullAgendaDto.events.map {it.toEntity()}
eventDao.insertEvents(eventEntities)
```

Reply:
@MKiperszmid updated